# Problem 1

In [1]:
threes = [x for x in range(3, 1000, 3)]
fives = [y for y in range (5, 999, 5)]
nums = list(set(fives+threes))
sum(nums)

233168

In [2]:
nums[:10]

[3, 5, 6, 9, 10, 12, 15, 18, 20, 21]

In [3]:
nums[-10:]

[980, 981, 984, 985, 987, 990, 993, 995, 996, 999]

# Problem 2

In [4]:
highest = 0
fib_list = [1, 1]
i = 0
while fib_list[-1] <= 4000000:
    fib_list.append(fib_list[i] + fib_list[i+1])
    i += 1
fib_list = fib_list[:-1]

even_sum = 0
for num in fib_list:
    if num % 2 == 0:
        even_sum += num

even_sum

4613732

# Problem 3

In [5]:
from functools import reduce

def factors(n):    
    return set(reduce(list.__add__, 
                ([i, n//i] for i in range(1, int(pow(n, 0.5) + 1)) if n % i == 0)))

In [6]:
factors_ = sorted(list(factors(600851475143)))
factors_

[1,
 71,
 839,
 1471,
 6857,
 59569,
 104441,
 486847,
 1234169,
 5753023,
 10086647,
 87625999,
 408464633,
 716151937,
 8462696833,
 600851475143]

In [7]:
idx = 3
print(factors_[idx])
600851475143 / factors_[idx]

1471


408464633.0

In [8]:
# is each factor in the list prime?
for each in factors_:
    print(factors(each))

{1}
{1, 71}
{1, 839}
{1, 1471}
{1, 6857}
{1, 59569, 839, 71}
{1, 1471, 104441, 71}
{1, 6857, 71, 486847}
{1, 1471, 1234169, 839}
{1, 6857, 839, 5753023}
{1, 6857, 1471, 10086647}
{1, 839, 71, 87625999, 59569, 1234169, 104441, 1471}
{1, 839, 71, 486847, 6857, 59569, 408464633, 5753023}
{1, 716151937, 71, 486847, 6857, 10086647, 104441, 1471}
{1, 8462696833, 5753023, 839, 6857, 10086647, 1234169, 1471}
{1, 8462696833, 716151937, 5753023, 1234169, 839, 71, 600851475143, 6857, 486847, 87625999, 59569, 10086647, 408464633, 104441, 1471}


From the above, we can see that the largest prime factor is 6857

# Problem 4

In [9]:
# The largest multiple would be 999 * 999, so let's start there.
palindrome_list = []
for i in range(900):
    first_num = 1000 - i
    for j in range(900):
        second_num = 1000 - j
        num_string = str(first_num * second_num)
        if num_string == num_string[::-1]:
            palindrome_list.append(first_num * second_num)
            

In [10]:
max(palindrome_list)

906609

# Problem 5

In [11]:
def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n - 1)

In [12]:
factorial(10)

3628800

In [13]:
factors_list = [list(factors(i)) for i in range(2, 21)]

prime_factors = []
for each in factors_list[1:]:
    each = sorted(each, reverse=True)
    done = False
    while not done:
        if sorted(list(factors(each[0]))) != [1, each[0]]:
            composite = each.pop(0)
            composite_factors = factors(composite)
            each.extend(list(composite_factors))
        else:
            prime_factors.append(each)
            done = True

# Numbers 1 through 20, reduced to their prime factorsb
prime_factors

[[3, 1],
 [2, 1, 1, 2, 4],
 [5, 1],
 [3, 2, 1, 1, 2, 3, 6],
 [7, 1],
 [2, 1, 8, 1, 2, 4, 1, 2, 4],
 [3, 1, 1, 3, 9],
 [5, 2, 1, 1, 10, 2, 5],
 [11, 1],
 [3, 2, 1, 1, 2, 3, 4, 6, 12, 1, 2, 3, 6, 1, 2, 4],
 [13, 1],
 [7, 2, 1, 1, 2, 14, 7],
 [5, 3, 1, 1, 3, 5, 15],
 [2, 1, 1, 2, 4, 8, 16, 8, 1, 2, 4, 1, 2, 4],
 [17, 1],
 [3, 2, 1, 1, 2, 3, 6, 9, 18, 1, 3, 9, 1, 2, 3, 6],
 [19, 1],
 [5, 4, 2, 1, 1, 2, 4, 5, 10, 20, 1, 10, 2, 5]]

I just did this one in my head

In [14]:
1*2*2*2*2*3*3*5*7*11*13*17*19

232792560

# Problem 6

In [15]:
sum_of_squares = sum([x**2 for x in range(1, 101)])
square_of_sums = sum([x for x in range(1, 101)]) ** 2
square_of_sums - sum_of_squares

25164150

# Problem 7

In [16]:
def is_prime(x):
    if sorted(list(factors(x))) == [1, x]:
        return True
    else:
        return False

is_prime(11)

True

In [17]:
import pickle
import numpy as np
import math

def nth_prime_number(n):
    try:
        list_of_primes = pickle.load(open('list_of_primes.pkl', 'rb'))
    except:
        # First 10 primes for free!
        list_of_primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
        pickle.dump(list_of_primes, open('list_of_primes.pkl', 'wb'))
    
    while len(list_of_primes) < n:
        old_length = len(list_of_primes)
        interval_extend = 100000
        interval_start = list_of_primes[-1] + 1
        interval_end = interval_start + interval_extend
        
        # Set up a boolean array for the new numbers to sift through
        bool_interval = np.ones(interval_extend, dtype=bool)
        for num in list_of_primes:
            # this sets up the bool index as the number it represents in the interval,
            # starting with the first multiple
            first_index = math.ceil((interval_start)/ num) * num - (interval_start)
            bool_interval[range(first_index, interval_extend, num)] = False
        
        original_list = list_of_primes.copy()
        
        # Money shot
        list_of_primes.extend(np.linspace(interval_start, 
                                          interval_end, 
                                          interval_extend, 
                                          dtype=int)[bool_interval].tolist())
        i = old_length
        while i < len(list_of_primes):
            if bool_interval[i - old_length]:
                p = list_of_primes[i]
#                 print(p, i)
                bool_interval[range(i-old_length+p-1, interval_extend, p)] = False
                temp_list = original_list.copy()
                
                list_of_primes = temp_list
                list_of_primes.extend(np.linspace(interval_start+1, 
                                                  interval_end, 
                                                  interval_extend, 
                                                  dtype=int)[bool_interval].tolist())
#                 print(i)
                
            i += 1 
        print(i)
    # Now that we have at least n primes in our list, return the nth prime
    return list_of_primes[n-1]
    
    
    
    
nth_prime_number(100)

15391


548

In [80]:
import pickle

def nth_prime(n):
    try:
        list_of_primes = pickle.load(open('list_of_primes.pkl', 'rb'))
    except:
        # First 10 primes for free!
        list_of_primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
        pickle.dump(list_of_primes, open('list_of_primes.pkl', 'wb'))


    while len(list_of_primes) < n:
        # Some variables we will need
        old_length = len(list_of_primes)
        interval_extend = 100000
        interval_start = list_of_primes[-1] + 1
        interval_end = interval_start + interval_extend

        # Set up a boolean array for the new numbers to sift through
        bool_interval = np.ones(interval_extend, dtype=bool)
        for num in list_of_primes:
            # this sets up the bool index as the number it represents in the interval,
            # starting with the first multiple
            first_index = math.ceil((interval_start)/ num) * num - (interval_start)
            bool_interval[range(first_index, interval_extend, num)] = False
            
        i = interval_start
        while i < np.sqrt(list_of_primes[-1] + interval_extend) + 2:
            if bool_interval[i - interval_start]:
                bool_interval[range(2*i - interval_start, interval_extend, i)] = False
            i += 1 

        list_of_primes.extend(np.linspace(interval_start, 
                                      interval_end, 
                                      interval_extend, 
                                      dtype=int)[bool_interval].tolist())
    # Now, save the list of primes in case we ever need them again
    pickle.dump(list_of_primes, open('list_of_primes.pkl', 'wb'))
    # Now that we have at least n primes in our list, return the nth prime
    return list_of_primes[n - 1]

nth_prime(10001)

104743

# Problem 8

In [99]:
import operator
# Let's save it as a string for indexing through it
big_ass_number = ('73167176531330624919225119674426574742355349194934'
                  '96983520312774506326239578318016984801869478851843'
                  '85861560789112949495459501737958331952853208805511'
                  '12540698747158523863050715693290963295227443043557'
                  '66896648950445244523161731856403098711121722383113'
                  '62229893423380308135336276614282806444486645238749'
                  '30358907296290491560440772390713810515859307960866'
                  '70172427121883998797908792274921901699720888093776'
                  '65727333001053367881220235421809751254540594752243'
                  '52584907711670556013604839586446706324415722155397'
                  '53697817977846174064955149290862569321978468622482'
                  '83972241375657056057490261407972968652414535100474'
                  '82166370484403199890008895243450658541227588666881'
                  '16427171479924442928230863465674813919123162824586'
                  '17866458359124566529476545682848912883142607690042'
                  '24219022671055626321111109370544217506941658960408'
                  '07198403850962455444362981230987879927244284909188'
                  '84580156166097919133875499200524063689912560717606'
                  '05886116467109405077541002256983155200055935729725'
                  '71636269561882670428252483600823257530420752963450')

products = []
for i in range(len(big_ass_number)-12):
    adj_nums = big_ass_number[i:i+13]
    adj_nums = [int(digit) for digit in adj_nums]
    adj_prod = (reduce(lambda x, y: x*y, adj_nums))
    products.append(adj_prod)
    
print(max(products))

23514624000


# Problem 9

In [106]:
# Here, we have two equations
# a + b + c = 1000
# and
# a^2 + b^2 + c^2
# We also have the constraint a < b < c
# The lowest value of c is 335
# Ehh. Just brute force it
for c in range(1, 1001):
    for b in range(1, 1000-c):
        a = 1000 - b - c
        if b > c:
            continue
        if a > b:
            continue
        if a**2 + b**2 == c**2:
            print(a*b*c)

31875000


Way easier :)

# Problem 10

In [203]:
# Let's use the list of prime numbers we saved in an earlier problem.
# Good thing we pickled it!
# Note, I had to increase the number of primes by running it a few times
primes_list = pickle.load(open('list_of_primes.pkl', 'rb'))

n = 100017
while primes_list[-1] < 2100000:
    print('Adding more numbers to primes')
    nth_prime(n)
    primes_list = pickle.load(open('list_of_primes.pkl', 'rb'))
    n += 100000

summation = 0
blah = []
for i, p in enumerate(primes_list):
    if p < 2000000:
        summation += p
        blah.append(p)
        if i < 5:
            print(primes_list[i], p)
    else:
        break

2 2
3 3
5 5
7 7
11 11


In [201]:
summation 

142913828925

In [188]:
sum(blah)

142913828925

In [189]:
sum(primes_list[:148933])

142913828925

The answer, according to Wolfram Alpha, is in the cell below. I am so confused right now. My prime number generator works up to like 200,000 prime numbers. I've checked multiple sites and they all give the same output for the nth_prime(n). Yet, somehow my answer is off by 3?!?!? I checked the beginning of the list and it is correct: the number 3 only appears once in my list_of_primes. So that means that somewhere in the list, a difference of 3 somehow worked its way into the numbers without messing up the rest of the numbers. Hmm.

In [198]:
142913828922 # The answer...

142913828922

In [196]:
summation - 142913828922

3

In [191]:
for each in list_of_primes:
    if sorted(list(factors(each))) != [1, each]:
        print(each)

This is infuriating. Every number in my list of primes is a prime number! Maybe there is a random 3 in the list somewhere? let's compare the length of the list to the length of the set of the list.

In [192]:
len(list_of_primes)

9594

In [209]:
len(list(set(list_of_primes)))

9594

In [217]:
list_of_primes == sorted(list(set(list_of_primes)))

True

In [194]:
list_of_primes == sorted(list_of_primes)

True

Yep. I give up. All numbers are unique, all numbers are primes, and the list does not change when it is sorted either. Why would my summation loop skip a 3? It's not even the first number!